FreeRTOS/Demo/CORTEX_MPU_M3_MPS2_QEMU_GCC/Readme.md

```diff
@@ -0,0 +1,93 @@
+# Emulating MPS2 Cortex M3 AN385 on QEMU
+
+## Requirements
+1. GNU Arm Embedded Toolchain download [here](https://developer.arm.com/tools-and-software/open-source-software/developer-tools/gnu-toolchain/gnu-rm/downloads) (tested on versiom 9.3.1 20200408)
+3. qemu-arm-system download [here](https://www.qemu.org/download) (tested on version 5.0.1 (v5.0.1-dirty))
+2. Make (tested on version 3.82)
+4. Linux OS (tested on Ubuntu 18.04)
+
+## How to download
+Navigate to a directory of your choice and run the following command
+```
+$ git clone https://github.com/FreeRTOS/FreeRTOS.git --recurse-submodules --depth 1
+```
+The previous command should create a directory named **FreeRTOS**
+
+## Getting Started on Windows using WSL
+The Windows Subsystem for Linux allows you to run native Linux applications from a shell on your windows machine.
+
+To set up your Windows 10 machine to run this QEMU based demo you can follow these steps
+1. Install Ubuntu 20.04 LTS version from Microsoft Store, search for "Ubuntu"
+2. Update apt-get
+```
+sudo apt-get update
+```
+3. Install Make and Qemu
+```
+sudo apt-get install -y make qemu qemu-system-arm
+```
+4. Download and unzip Arm tools
+```
+cd ~
+curl https://armkeil.blob.core.windows.net/developer/Files/downloads/gnu-rm/10-2020q4/gcc-arm-none-eabi-10-2020-q4-major-x86_64-linux.tar.bz2 -o gcc-arm-none-eabi-10-2020-q4-major-x86_64-linux.tar.bz2
+tar -xjvf gcc-arm-none-eabi-10-2020-q4-major-x86_64-linux.tar.bz2
+```
+
+5. Update your path to include the arm toolchain Edit ".profile", add the unzipped bin folder to the front of the path. You can run the same command in the terminal to update the path temporarily in the current shell
+```
+export PATH="$HOME/gcc-arm-none-eabi-10-2020-q4-major/bin:$PATH"
+```
+
+6. Clone FreeRTOS
+```
+git clone https://github.com/FreeRTOS/FreeRTOS.git --recurse-submodules
+```
+
+7. Compile the code
+```
+$ cd ./FreeRTOS/FreeRTOS/Demo/CORTEX_MPU_M3_MPS2_QEMU_GCC
+$ export PATH=/path/to/arm/toolchain:$PATH
+$ make
+```
+a file named __./build/RTOSDemo.axf__ will be generated
+
+8. Run the MPU Demo
+```
+sudo qemu-system-arm -machine mps2-an385 -monitor null -semihosting --semihosting-config enable=on,target=native -kernel ./build/RTOSDemo.axf -serial stdio -nographic -s -S
+```
+__ -s -S __ will start the program waiting for the debugger, check the next
+section on how to attack the debuffer to the program and observe how it it not
+causing memory faults by placing break points on the MemMang\_handler in __startup.c__
+and inside the function __prvRWAccessTask__ in mpu\_demo.c
+To be noticed is how the task __prvRWAccessTask__ does not cause a memory fault
+while the __prvROAccessTask__ causes a memory violation with the MPU
+
+For a versions with debugging symbols and no optimizations **-O0**, run:
+```
+$ make DEBUG=1
```

Comment:
Move it above where is build command is. Should it not be default as we ask the user to attach a debugger?

Reply:
they can just use printing if they want, it works in qemu